src/panels/lovelace/components/hui-energy-period-selector.ts

```diff
@@ -128,172 +203,153 @@ export class HuiEnergyPeriodSelector extends SubscribeMixin(LitElement) {
+  private _dateRangeChanged(ev) {
     const weekStartsOn = firstWeekdayIndex(this.hass.locale);
-
-    this._setDate(
-      this._period === "day"
-        ? calcDate(new Date(), startOfDay, this.hass.locale, this.hass.config)
-        : this._period === "week"
-        ? calcDate(
-            new Date(),
-            startOfWeek,
-            this.hass.locale,
-            this.hass.config,
-            {
-              weekStartsOn,
-            }
-          )
-        : this._period === "month"
-        ? calcDate(new Date(), startOfMonth, this.hass.locale, this.hass.config)
-        : calcDate(new Date(), startOfYear, this.hass.locale, this.hass.config)
+    this._startDate = calcDate(
+      ev.detail.startDate,
+      startOfDay,
+      this.hass.locale,
+      this.hass.config,
+      {
+        weekStartsOn,
+      }
+    );
+    this._endDate = calcDate(
+      ev.detail.endDate,
+      endOfDay,
+      this.hass.locale,
+      this.hass.config,
+      {
+        weekStartsOn,
+      }
     );
+
+    this._updateCollectionPeriod();
   }
 
   private _pickPrevious() {
-    const newStart =
-      this._period === "day"
-        ? addDays(this._startDate!, -1)
-        : this._period === "week"
-        ? addWeeks(this._startDate!, -1)
-        : this._period === "month"
-        ? addMonths(this._startDate!, -1)
-        : addYears(this._startDate!, -1);
-    this._setDate(newStart);
+    this._shift(false);
   }
 
   private _pickNext() {
-    const newStart =
-      this._period === "day"
-        ? addDays(this._startDate!, 1)
-        : this._period === "week"
-        ? addWeeks(this._startDate!, 1)
-        : this._period === "month"
-        ? addMonths(this._startDate!, 1)
-        : addYears(this._startDate!, 1);
-    this._setDate(newStart);
+    this._shift(true);
   }
 
-  private _setDate(startDate: Date) {
-    const weekStartsOn = firstWeekdayIndex(this.hass.locale);
+  private _shift(forward: boolean) {
+    if (!this._startDate) return;
 
-    const endDate =
-      this._period === "day"
-        ? calcDate(startDate, endOfDay, this.hass.locale, this.hass.config)
-        : this._period === "week"
-        ? calcDate(startDate, endOfWeek, this.hass.locale, this.hass.config, {
-            weekStartsOn,
-          })
-        : this._period === "month"
-        ? calcDate(startDate, endOfMonth, this.hass.locale, this.hass.config)
-        : calcDate(startDate, endOfYear, this.hass.locale, this.hass.config);
+    let start: Date;
+    let end: Date;
+    if (
+      isFirstDayOfMonth(this._startDate) &&
+      isLastDayOfMonth(this._endDate!)
+    ) {
+      // Shift date range with respect to month/year selection
+      const difference =
+        (differenceInMonths(this._endDate!, this._startDate) + 1) *
+        (forward ? 1 : -1);
+      start = addMonths(this._startDate, difference);
+      end = endOfMonth(addMonths(this._endDate!, difference));
+    } else {
+      // Shift date range by period length
```

Comment:
A quarter and a year also don't have a fixed amount of days, so should get the same treatment as months.

Reply:
I think no change is necessary since both Quarters and Years start/end with the first/last day of the Month. The condition covers these cases and all other multiple-month selections.